octomap_mapping/octomap_server/src/zFilter.cpp

```diff
@@ -37,47 +42,87 @@ void trimmed_cb(const octomap_msgs::Octomap& input){ // occupancy tree call back
 		}
 	}
 
+	
 	cloudOccFull->header.frame_id = "/world";
 	cloudOccFull->width = sizeOfCloud; cloudOccFull->height = 1; cloudOccFull->points.resize (cloudOccFull->width * cloudOccFull->height);
 	for(it = trimmedOcTree->begin_leafs(),endLeaf = trimmedOcTree->end_leafs();it!=endLeaf;++it){
 		if(it->getValue()>thresholdOcc){
 			if(it.getZ()>trimmedHeight){
 				cloudOccFull->points[id4Markers].x = it.getX(); cloudOccFull->points[id4Markers].y = it.getY(); cloudOccFull->points[id4Markers].z = it.getZ();
+				id4Markers++;
+			}
+		}
+	}
+
+	pcl::KdTreeFLANN<pcl::PointXYZ> kdtree;
 
-				zFilteredMarkers.pose.position.x = it.getX(); zFilteredMarkers.pose.position.y = it.getY(); zFilteredMarkers.pose.position.z = it.getZ();
-				zFilteredMarkers.id = id4Markers;
+	kdtree.setInputCloud (cloudOccFull);
+	// K nearest neighbor search
+	int radius = 1;
+	// int K = 2;
+	std::vector<int> pointIdxNKNSearch;
+	std::vector<float> pointNKNSquaredDistance;
+
+	visualization_msgs::Marker zPrunedMarkers; zPrunedMarkers.header.frame_id = "/world"; zPrunedMarkers.header.stamp = ros::Time::now(); zPrunedMarkers.ns = "free_markers_pruned";
+	zPrunedMarkers.type = shape; zPrunedMarkers.action = visualization_msgs::Marker::ADD;
+	zPrunedMarkers.color.r = 1.0f; zPrunedMarkers.color.g = 0.0f; zPrunedMarkers.color.b = 0.0f; zPrunedMarkers.color.a = 1.0;
+
+
```

Comment:
This loop searches for neighbors for each of the point in the octomap point cloud. If the no. of neighbors are greater than 'initstraySizeThresh', the point is taken as a valid usable point.